source/fuzz/fuzzer_pass_replace_irrelevant_ids.cpp

```diff
@@ -0,0 +1,160 @@
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassReplaceIrrelevantIds::~FuzzerPassReplaceIrrelevantIds() = default;
+
+void FuzzerPassReplaceIrrelevantIds::Apply() {
+  // Keep track of the irrelevant ids.
+  std::vector<uint32_t> irrelevant_ids;
+
+  // Keep a map from the type ids of irrelevant ids to all the ids with that
+  // type.
+  std::unordered_map<uint32_t, std::vector<uint32_t>> types_to_ids;
+
+  for (auto id :
+       GetTransformationContext()->GetFactManager()->GetIrrelevantIds()) {
+    // Check that the id still exists in the module.
+    auto declaration = GetIRContext()->get_def_use_mgr()->GetDef(id);
+    if (!declaration) {
+      continue;
+    }
+
+    irrelevant_ids.push_back(id);
+
+    // If the type of this id has not been seen before, add a mapping from this
+    // type id to an empty list in |types_to_ids|. The list will be filled later
+    // on.
+    if (types_to_ids.count(declaration->type_id()) == 0) {
+      types_to_ids[declaration->type_id()] = {};
+    }
+  }
+
+  // If no irrelevant ids were found, return.
+  if (irrelevant_ids.empty()) {
+    return;
+  }
+
+  // Loop through all the ids in the module and fill all the empty lists in
+  // |types_to_ids|.
+  for (const auto& pair : GetIRContext()->get_def_use_mgr()->id_to_defs()) {
+    uint32_t type_id = pair.second->type_id();
+    if (types_to_ids.count(type_id)) {
+      types_to_ids[type_id].push_back(pair.first);
+    }
+  }
+
+  // Keep a list of all the transformations to perform. We avoid applying the
+  // transformations while traversing the uses since applying the transformation
+  // invalidates all analyses, and we want to avoid invalidating and recomputing
+  // them every time.
+  std::vector<TransformationReplaceIrrelevantId> transformations_to_apply;
+
+  // Loop through all the uses of irrelevant ids, check that the id can be
+  // replaced and randomly decide whether to apply the transformation.
+  for (auto irrelevant_id : irrelevant_ids) {
+    uint32_t type_id =
+        GetIRContext()->get_def_use_mgr()->GetDef(irrelevant_id)->type_id();
+
+    GetIRContext()->get_def_use_mgr()->ForEachUse(
+        irrelevant_id, [this, &irrelevant_id, &type_id, &types_to_ids,
+                        &transformations_to_apply](opt::Instruction* use_inst,
+                                                   uint32_t use_index) {
+          // The id must be used as an input operand.
+          if (use_index < use_inst->NumOperands() - use_inst->NumInOperands()) {
+            // The id is used as an output operand, so we cannot replace this
+            // usage.
+            return;
+          }
+
+          // Randomly decide whether to consider this use.
+          if (!GetFuzzerContext()->ChoosePercentage(
+                  GetFuzzerContext()->GetChanceOfReplacingIrrelevantId())) {
+            return;
+          }
```

Comment:
We tend to do this check first - probabilistically I don't think there's any disadvantage to doing so.  I.e., I'd move it to be the first thing that happens in the lambda.

Reply:
Done